core/src/transport/circuit_breaking.rs

```diff
@@ -0,0 +1,383 @@
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+/// Timeout set when opening CircuitBreaker.
+const INITIAL_TIMEOUT: Duration = Duration::from_secs(2);
+/// Maximum allowed timeout.
+const MAX_TIMEOUT: Duration = Duration::from_secs(300);
+
+/// Wraps a [`Transport`] and implements a [circuit breaker
+/// pattern](https://docs.microsoft.com/en-us/azure/architecture/patterns/circuit-breaker)
+/// for addresses that cannot be dialed. The circuit breaker pattern aims to
+/// save resources by delaying an operation which is likely to fail.
+///
+/// If [`Transport::dial`] fails, a circuit is opened for the dialed address
+/// `addr`. When dialing `addr` again while the circuit is open,
+/// [`CircuitBreakingDial`] adds a delay before calling `dial` on the underlying
+/// `Transport`. The delay for `addr` increases if dialing it fails repeatedly.
+///
+/// Open circuits are closed automatically after some time. In addition, an open
+/// circuit for an address is closed after successfully dialing that address.
+#[derive(Clone)]
+pub struct CircuitBreaking<TInner> {
+    /// The underlying transport.
+    inner: TInner,
+    state: Arc<CircuitBreakerState>,
+}
+
+/// Represents open circuits.
+struct CircuitBreakerState {
+    addresses: Mutex<FnvHashMap<Multiaddr, AddrInfo>>,
+}
+
+#[derive(Debug)]
+struct AddrInfo {
+    /// The delay to add before calling [`Transport::dial`].
+    wait_duration: Duration,
+    /// After `expiration` the circuit can be closed.
+    expiration: Delay,
+}
+
+impl CircuitBreakerState {
+    fn new() -> Self {
+        CircuitBreakerState {
+            addresses: Mutex::new(Default::default()),
+        }
+    }
+
+    /// Returns `None` if circuit of `addr` is closed.
+    fn get_wait_duration(&self, addr: &Multiaddr) -> Option<Duration> {
+        self.addresses
+            .lock()
+            .get(addr)
+            .map(|addr_info| addr_info.wait_duration)
+    }
+
+    fn open_or_extend_circuit(&self, addr: Multiaddr) {
+        match self.addresses.lock().entry(addr) {
+            Entry::Vacant(entry) => {
+                entry.insert(AddrInfo {
+                    wait_duration: INITIAL_TIMEOUT,
+                    expiration: Delay::new(INITIAL_TIMEOUT * 2),
+                });
+            }
+
+            Entry::Occupied(entry) => {
+                let addr_info = entry.into_mut();
+                addr_info.wait_duration = cmp::min(addr_info.wait_duration * 2, MAX_TIMEOUT);
+                addr_info.expiration = Delay::new(addr_info.wait_duration * 2);
+            }
+        }
+    }
+
+    /// Closes the circuit of `addr` even if the circuit has not yet expired.
+    fn close_circuit(&self, addr: &Multiaddr) {
+        self.addresses.lock().remove(addr);
+    }
+
+    fn close_expired_circuits(&self) {
+        let mut addresses = self.addresses.lock();
+        addresses.retain(move |_, a| (&mut a.expiration).now_or_never().is_none());
```

Comment:
If I am not mistaken, `AddrInfo::expiration` could as well be a `std::time::Instant` and a simple comparison here (`a.expiration < now`) would suffice, right?

Reply:
Done in 101a36d.